src/dynsight/_internal/vision/detect.py

```diff
@@ -0,0 +1,782 @@
+                            "width": float(w),
+                            "height": float(h),
+                            "confidence": float(conf[i]),
+                        }
+                    )
+        # Initialize outliers folder in the prediction folder
+        outliers_plt_folder = (
+            self.project_folder
+            / "predictions"
+            / f"attempt_{prediction_number}"
+            / "outliers"
+        )
+        outliers_plt_folder.mkdir(exist_ok=True)
+
+        # Look for outliers in the boxes width and height
+        widths = np.array([d["width"] for d in detection_results], dtype=float)
+        heights = np.array(
+            [d["height"] for d in detection_results],
+            dtype=float,
+        )
+        out_width = set(
+            _find_outliers(
+                distribution=widths,
+                save_path=outliers_plt_folder,
+                fig_name="width",
+            )
+        )
+        out_height = set(
+            _find_outliers(
+                distribution=heights,
+                save_path=outliers_plt_folder,
+                fig_name="height",
+            )
+        )
+        # Exclude the outliers from the detection results
+        filtered_detections = [
+            det
+            for det in detection_results
+            if (det["width"] not in out_width)
+            and (det["height"] not in out_height)
+        ]
+        detection_results = filtered_detections
+
+        # Build a new dataset based on the "filtered" detection results
+        # New dataset path
+        train_dataset_path = (
+            self.project_folder
+            / "train_datasets"
+            / f"dataset_{prediction_number}"
+        )
+
+        # Build the dataset
+        self._build_dataset(
+            detection_results=detection_results,
+            dataset_name=f"dataset_{prediction_number}",
+        )
+        # Add the new dataset to the training config file
+        self._add_or_create_yaml(train_dataset_path)
+
+        # Iterative part to improve the model performace
+        for s in range(max_sessions):
+            new_model_name = f"v{s + 1}"
+            self.train(
+                yaml_file=self.yaml_file,
+                initial_model=current_model_path,
+                training_epochs=training_epochs,
+                training_patience=training_patience,
+                batch_size=batch_size,
+                workers=workers,
+                device=device,
+                training_name=new_model_name,
+            )
+            current_model_path = (
+                self.project_folder
+                / "models"
+                / new_model_name
+                / "weights"
+                / "best.pt"
+            )
+            current_model = YOLO(current_model_path)
+            prediction_number += 1
+            detection_results = []
+            for f in range(0, self.n_frames, 50):  # TEMP
```

Comment:
TEMP?

Also, see the duplicate code here and in the first generation? Possible to make each of these loops a method?